### ISSUE TITLE
Kubelet service on wrong port

### ISSUE THREAD

**Issue by ashiddo11** [prometheus-operator/prometheus-operator#1400]
I'm using the prometheus operator chart with kubernetes 1.9.6
The kubelet service is listening on port 10250 however cadvisor on the nodes listens on 10255. Can we make the kubelet port a variable so that we can change it depending on the port we're using?

**Comment by brancz** [comment]
Actually the underlying endpoints object already genrates all 3 ports:

```
  ports:
  - name: http-metrics
    port: 10255
    protocol: TCP
  - name: cadvisor
    port: 4194
    protocol: TCP
  - name: https-metrics
    port: 10250
    protocol: TCP
```

It's a bug, that the Service doesn't have all three.

**Comment by ashiddo11** [comment]
It works if I change port from **https-metrics** to **http-metrics** even thought the service doesn't have all three ports

```apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  clusterName: ""
  labels:
    prometheus: main
  name: kubelet
  namespace: kube-system
spec:
  endpoints:
  - interval: 30s
    port: http-metrics
  jobLabel: kubelet
  namespaceSelector:
    matchNames:
    - kube-system
  selector:
    matchLabels:
      k8s-app: kubelet
```

```
apiVersion: v1
kind: Service
metadata:
  labels:
    k8s-app: kubelet
  name: kubelet
  namespace: kube-system
spec:
  clusterIP: None
  ports:
  - name: https-metrics
    port: 10250
    protocol: TCP
    targetPort: 10250
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

**Comment by ashiddo11** [comment]
Im happy to open a PR if you can point me in the right direction

**Comment by brancz** [comment]
The reconciliation of nodes starts here: https://github.com/coreos/prometheus-operator/blob/254edd18d54e4eb9c88ddb3a7975e2c792908676/pkg/prometheus/operator.go#L368

And concretely the service port definition is here: https://github.com/coreos/prometheus-operator/blob/254edd18d54e4eb9c88ddb3a7975e2c792908676/pkg/prometheus/operator.go#L468

Make sure naming and port numbers are consistent with this: https://github.com/coreos/prometheus-operator/blob/254edd18d54e4eb9c88ddb3a7975e2c792908676/pkg/prometheus/operator.go#L419-L433

Thanks a lot for tackling this! :slightly_smiling_face:

**Comment by raftAtGit** [comment]
This is Azure AKS with Kube 1.9.6.

In my case switching to http `--set exporter-kubelets.https=false` didnt solve my problem. Turned out to be exporter was trying to connect `cadvisor` port.

Fixed by modifying `templates/servicemonitor.yaml` from 
```
{{- else }}
  - port: http-metrics
    interval: 15s
  - port: cadvisor
    interval: 30s
    honorLabels: true
{{- end }}
```
to 

```
{{- else }}
  - port: http-metrics
    interval: 15s
  - port: http-metrics
    path: /metrics/cadvisor
    interval: 30s
    honorLabels: true
{{- end }}
```

**Comment by aneeshwara** [comment]
hi i am trying to kubeadm init phase preflight
 unable to initiate and i am getting error as 10250 port is in use

**Comment by aneeshwara** [comment]
and 10250 port in use by kubelet

**Comment by cpick** [comment]
Still affected, so commenting to remove "stale".

**Comment by cpick** [comment]
Commenting to remove "stale".

**Comment by yeya24** [comment]
Is this issue still valid? This work well for me

**Comment by sschne** [comment]
Looks like this has been fixed with #3039

**Comment by simonpasquier** [comment]
Thanks! Closing it then :)
